File: Prime_numbers_factors_&_number_theory/Q3. Closest Prime Numbers in Range/Q3_ClosestPrimeNumbersinRange.py
class Solution(object):
    def closestPrimes(self, left, right):
        """
        :type left: int
        :type right: int
        :rtype: List[int]
        """
        
        limit = right
        # We need primes only up to 'right'

        isPrime = [True] * (limit + 1)
        # Create an array marking all numbers as prime initially

        if limit >= 0:
            isPrime[0] = False
        # 0 is not a prime

        if limit >= 1:
            isPrime[1] = False
        # 1 is not a prime

        p = 2
        # Start checking primes from 2

        while p * p <= limit:
            # Loop until square root of limit

            if isPrime[p]:
                # If number is prime

                for multiple in range(p * p, limit + 1, p):
                    # Mark all multiples of p as not prime

                    isPrime[multiple] = False

            p += 1
            # Move to next number

        primes = []
        # List to store primes in the given range

        for i in range(left, right + 1):
            # Check numbers from left to right

            if isPrime[i]:
                primes.append(i)
                # Add prime number to list

        if len(primes) < 2:
            # If less than two primes, no pair exists

            return [-1, -1]

        minDiff = float("inf")
        # Store smallest difference found so far

        ans = [-1, -1]
        # Store final answer pair

        for i in range(len(primes) - 1):
            # Loop through consecutive primes

            diff = primes[i + 1] - primes[i]
            # Find difference between pair

            if diff < minDiff:
                # Found a smaller difference

                minDiff = diff
                ans = [primes[i], primes[i + 1]]
                # Update answer pair

        return ans
        # Return closest prime pair

File: Prime_numbers_factors_&_number_theory/Q3. Closest Prime Numbers in Range/Q3_ClosestPrimeNumbersinRange.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    public int[] closestPrimes(int left, int right) {

        int limit = Math.max(right, 1000000);
        // Take max range so sieve can cover at least 'right'

        boolean[] isPrime = new boolean[limit + 1];
        // Boolean array to mark prime numbers

        Arrays.fill(isPrime, true);
        // Initially assume all numbers are prime

        isPrime[0] = false;
        isPrime[1] = false;
        // 0 and 1 are not primes

        for (int num = 2; num * num <= limit; num++) {
            // Loop through numbers up to sqrt(limit)

            if (isPrime[num]) {
                // If number is still marked prime

                for (int multiple = num * num; multiple <= limit; multiple += num) {
                    // Mark all multiples as not prime

                    isPrime[multiple] = false;
                }
            }
        }

        List<Integer> primes = new ArrayList<>();
        // List to store primes in the range

        for (int i = left; i <= right; i++) {
            // Loop through the range [left, right]

            if (isPrime[i]) {
                primes.add(i);
                // Add prime to list
            }
        }

        if (primes.size() < 2) {
            // If less than two primes exist, no pair possible

            return new int[]{-1, -1};
        }

        int minDiff = Integer.MAX_VALUE;
        // Track smallest difference between primes

        int[] result = new int[2];
        // Store closest prime pair

        for (int i = 0; i < primes.size() - 1; i++) {
            // Check each consecutive pair

            int diff = primes.get(i + 1) - primes.get(i);
            // Difference between pair

            if (diff < minDiff) {
                // If we found a smaller difference

                minDiff = diff;
                result[0] = primes.get(i);
                result[1] = primes.get(i + 1);
                // Update closest pair
            }
        }

        return result;
        // Return final closest pair
    }
}
